pkg/typesystem/typesystem.go

```diff
@@ -257,6 +280,28 @@ func validateRelationTypeRestrictions(model *openfgapb.AuthorizationModel) error
 					}
 				}
 			}
+			// forbid the following model because `viewer` is a cycle
+			// type folder
+			//	 relations
+			//		define parent: [folder] as self
+			//		define viewer as viewer from parent
+
+			switch cyclicDefinition := relation.GetRewrite().Userset.(type) {
```

Comment:
Should this be part of `validateRelationTypeRestrictions`? Perhaps it can be moved to its own function, such as `ensureNoCycles`, that we can add to later with other cases as well?